pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,790 @@
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+	}
+
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		if errors.IsNotFound(err) {
+			log.Infof("k8s secret not found during delete. Skip. secret: %s, ns: %s", name, namespace)
+			return nil
+		}
+		log.Error(err, "error while getting K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	if err := c.Delete(ctx, secret); err != nil {
+		log.Error(err, "error while deleting K8s secret: %s, ns: %s", name, namespace)
+		return err
+	}
+
+	log.Infof("deleted k8s secret: %s, ns: %s", name, namespace)
+	return nil
+}
+
+// setStatus adds pod-specific info to byPod status of the secretproviderclass object
+func setStatus(ctx context.Context, obj *unstructured.Unstructured, id string, namespace string) error {
+	log.Infof("setStatus for pod: %s, ns: %s", id, namespace)
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	status := map[string]interface{}{
+		"id":        id,
+		"namespace": namespace,
+	}
+	statuses, exists, err := unstructured.NestedSlice(obj.Object, "status", "byPod")
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if err := unstructured.SetNestedSlice(
+			obj.Object, []interface{}{status}, "status", "byPod"); err != nil {
+			return err
+		}
+	}
+
+	for _, s := range statuses {
+		curStatus, ok := s.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		curID2, ok := curStatus["id"]
+		if !ok {
+			continue
+		}
+		curID, ok := curID2.(string)
+		if !ok {
+			continue
+		}
+		// skip if id already exists
+		if id == curID {
+			return nil
+		}
+	}
+
+	statuses = append(statuses, status)
+	if err := unstructured.SetNestedSlice(
+		obj.Object, statuses, "status", "byPod"); err != nil {
+		return err
+	}
+	err = c.Update(ctx, obj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// deleteStatus deletes pod-specific information from byPod status of the secretproviderclass object
+func deleteStatus(ctx context.Context, obj *unstructured.Unstructured, id string) error {
+	// recreating client here to prevent reading from cache
+	c, err := getClient()
+	if err != nil {
+		return err
+	}
+	statuses, exists, err := unstructured.NestedSlice(obj.Object, "status", "byPod")
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return nil
+	}
+	newStatus := make([]interface{}, 0)
+
+	for i, s := range statuses {
+		curStatus, ok := s.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("element %d in byPod status is malformed for pod: %s", i, id)
+		}
+		curID2, ok := curStatus["id"]
+		if !ok {
+			return fmt.Errorf("element %d in byPod status is missing an `id` field for pod: %s", i, id)
+		}
+		curID, ok := curID2.(string)
+		if !ok {
+			return fmt.Errorf("element %d in byPod status' `id` field is not a string: %v for pod: %s", i, curID2, id)
+		}
+		if id == curID {
+			continue
+		}
+		newStatus = append(newStatus, s)
+	}
```

Comment:
What if the pod UID never exists in the `status.byPod`? In that case we don't want to make an update secretProviderClass call as the UID doesn't exist.